pkg/k8splan/watcher.go

```diff
@@ -0,0 +1,196 @@
+	w := &watcher{
+		connInfo:    connInfo,
+		applyinator: applyinator,
+	}
+
+	go w.start(ctx)
+
+	return nil
+}
+
+type watcher struct {
+	connInfo                   config.ConnectionInfo
+	applyinator                applyinator.Applyinator
+	lastAppliedResourceVersion string
+}
+
+func (w *watcher) start(ctx context.Context) {
+	kc, err := clientcmd.RESTConfigFromKubeConfig([]byte(w.connInfo.KubeConfig))
+	if err != nil {
+		panic(err)
+	}
+
+	clientFactory, err := client.NewSharedClientFactory(kc, nil)
+	if err != nil {
+		panic(err)
+	}
+
+	cacheFactory := cache.NewSharedCachedFactory(clientFactory, &cache.SharedCacheFactoryOptions{
+		DefaultNamespace: w.connInfo.Namespace,
+		DefaultTweakList: func(options *metav1.ListOptions) {
+			options.FieldSelector = fmt.Sprintf("metadata.name=%s", w.connInfo.SecretName)
+		},
+	})
+
+	controllerFactory := controller.NewSharedControllerFactory(cacheFactory, nil)
+	core := corecontrollers.New(controllerFactory)
+
+	probePeriod, err := time.ParseDuration(enqueueAfterDuration)
+	if err != nil {
+		panic(err)
+	}
+
+	core.Secret().OnChange(ctx, "secret-watch", func(s string, secret *v1.Secret) (*v1.Secret, error) {
+		if rawPeriod, ok := secret.Data[probePeriodKey]; ok {
+			if parsedPeriod, err := time.ParseDuration(fmt.Sprintf("%ss", string(rawPeriod))); err != nil {
+				logrus.Errorf("[K8s] error parsing duration %ss, using default", string(rawPeriod))
+			} else {
+				probePeriod = parsedPeriod
+			}
+		}
+		if secret == nil {
+			logrus.Debugf("[K8s] Secret was nil")
+			core.Secret().EnqueueAfter(w.connInfo.Namespace, w.connInfo.SecretName, probePeriod)
+			return secret, nil
+		}
+		secret = secret.DeepCopy()
+		logrus.Debugf("[K8s] Processing secret %s in namespace %s at generation %d with resource version %s", secret.Name, secret.Namespace, secret.Generation, secret.ResourceVersion)
+		if w.lastAppliedResourceVersion == secret.ResourceVersion {
+			logrus.Errorf("last applied resource version (%s) did not change. skipping apply.", w.lastAppliedResourceVersion)
+			core.Secret().EnqueueAfter(w.connInfo.Namespace, w.connInfo.SecretName, probePeriod)
+			return secret, nil
+		}
+		if planData, ok := secret.Data[planKey]; ok {
+			logrus.Tracef("[K8s] Byte data: %v", planData)
+			logrus.Tracef("[K8s] Plan string was %s", string(planData))
+
+			var probeStatuses map[string]prober.ProbeStatus
+			// retrieve existing probe statuses from the secret if they exist
+			if rawProbeStatusByteData, ok := secret.Data[probeStatusesKey]; ok {
+				if err := json.Unmarshal(rawProbeStatusByteData, &probeStatuses); err != nil {
+					logrus.Errorf("[K8s] error while parsing probe statuses: %v", err)
+					probeStatuses = make(map[string]prober.ProbeStatus, 0)
+				}
+			}
+			// calculate the checksum of the plan from the provided data
+			cp, err := applyinator.CalculatePlan(planData)
+			if err != nil {
+				return secret, err
```

Comment:
Should this return nil for the secret since there's an error?

Reply:
the convention is to return the object passed into the `OnChange`, so I don't believe we should be returning `nil`